README.md
## Dial Gauge Web Component
[![Published on npm](https://img.shields.io/badge/npm-published-blue)](https://www.npmjs.com/package/dial-gauge)
[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/dial-gauge)
[![Version](https://img.shields.io/badge/version-0.1.5-brightgreen)](https://github.com/smontanus/dial-gauge)

Dial Gauge is a web component that provides a customizable dial gauge for numeric display. It is suitable for progress indication as well as graphical display of real time numeric data. It utilizes SVG and is written in pure Javascript with no dependencies.

### Installation
* **NPM**
    `npm install dial-gauge`

* **CDN**
    ```<script type="module" src="https://www.unpkg.com/dial-gauge/dial-gauge.js"></script>```

### Usage
The Dial Gauge component utilizes the custom element tag `<dial-gauge>`. The custom element has the following attributes:

* value - The custom element component value.
* scale-start - The custom element scale starting value.
* scale-end - The custom element scale ending value.
* main-title - The custom element main title, displayed at the top of the component.
* sub-title - The custom element subtitle, displayed at the bottom of the component.
* scale-offset - The custom element bottom offset, an angle applied to adjust the beginning and end of the scale.

The component will automatically update when attributes are set dynamically and displays a numeric representation of the data, as well as the graphical dial. Values that are outside the scale range will display OL for outside limits. The component 'sub-title' attribute is suitable for displaying the units of the data.

```html
<dial-gauge main-title="Main" sub-title="Sub" scale-start="0" scale-end="100" scale-offset="30" value="50">
</dial-gauge>
```

### CSS Style Variables
The following CSS property variables, along with their default values, are available for custom styling.
|Property Variable|Default Value|
|---|---|
|***General Element***|
|--dg-background-color|#ffffff|
|--dg-border|1px solid #000000|
|--dg-border-radius|5px|
|--dg-box-shadow|3px 3px 5px 1px rgba(0,0,0,0.35)|
|--dg-display|inline-block|
|--dg-height|200px|
|--dg-margin|0|
|--dg-width|200px|
|--dg-z-index|0|
|***Title***|
|--dg-title-color|#000000|
|--dg-title-font-family|Verdana, Geneva, Tahoma, sans-serif|
|--dg-title-font-size|1.5em|
|***Dial Arc***|
|--dg-arc-background-color|#efefef|
|--dg-arc-color|#000000|
|--dg-arc-width|20|
|***Numeric Display***|
|--dg-numeric-color|#000000|
|--dg-numeric-font-family|Verdana, Geneva, Tahoma, sans-serif|
|--dg-numeric-font-size|2.5em|
|***Subtitle***|
|--dg-subtitle-color|#000000|
|--dg-subtitle-font-family|Verdana, Geneva, Tahoma, sans-serif|
|--dg-subtitle-font-size|1em|

### Changing the Tag Name
The Dial Gauge component uses the `<dial-gauge>` tag in html. In certain cases users may want, or need, to change the tag name. This can be accomplished in Javascript by subclassing the element and registering this subclass under a new tag name.

```javascript
import {DialGauge} from 'dial-gauge';

customElements.define('my-new-tag', class extends DialGauge{});
```

### Contributing
1. Fork it!
2. Create your feature branch: `git checkout -b my-new-feature`
3. Commit your changes: `git commit -am 'Add some feature'`
4. Push to the branch: `git push origin my-new-feature`
5. Submit a pull request.


CHANGES.md
Dial Gauge Changelog
--------------
#### Version 0.1.5 (2023-12-15)
- Fixed typo in package.json.[@jogibear9988]

#### Version 0.1.4 (2023-12-10)
- Added :host() pseudo-class for direct access to shadow DOM CSS.[@jogibear9988]
- Implemented an AdoptedStylesheet.[@jogibear9988]
- Added custom-element.json file reference to package.json.[@jogibear9988]


#### Version 0.1.3 (2023-11-08)
- Added scale offset attribute.
- Fixed value out of scale limit when scale changed.


#### Version 0.1.2 (2023-03-04)

- Fixed bug where repeating values were not displayed.
- Added handling of values outside the scale limits.


#### Version 0.1.1 (2023-02-17)

- First public release.


[@smontanus]: https://github.com/smontanus


dial-gauge.js
/**
 * @file dial-gauge.js
 * @fileOverview Dial Gauge is a web component that provides a
 * customizable dial gauge for numeric display. It utilizes SVG
 * and is written in pure Javascript with no dependencies.
 * @version 0.1.5
 * @author Stephen Montanus <steve@stevemontanus.com>
 * @copyright ©2023 Stephen Montanus Software Engineering.
 * @license MIT
 */

/** @constant {Object} */
const template = document.createElement('template');
template.innerHTML = `
<div class="dial-gauge">
    <svg width="100%" height="100%">
        <text id="gauge-title" x="50%" y="11%"></text>
        <path id="gauge-background-arc" fill="none" />
        <path id="gauge-arc" fill="none" />
        <text id="gauge-numeric" x="50%" y="50%">0.0</text>
        <text id="gauge-subtitle" x="50%" y="96%"></text>
    </svg>
</div>
`;

const styleSheet = new CSSStyleSheet();
styleSheet.replaceSync(`
    :host {
        /* Element */
        --dg-background-color: #ffffff;
        --dg-border: 1px solid #000000;
        --dg-border-radius: 5px;
        --dg-box-shadow: 3px 3px 5px 1px rgba(0,0,0,0.35);
        --dg-display: inline-block;
        --dg-height: 200px;
        --dg-margin: 0;
        --dg-width: 200px;
        --dg-z-index: 0;
    
        /* Title */
        --dg-title-color: #000000;
        --dg-title-font-family: Verdana, Geneva, Tahoma, sans-serif;
        --dg-title-font-size: 1.5em;
    
        /* Dial Arc */
        --dg-arc-background-color: #efefef;
        --dg-arc-color: #000000;
        --dg-arc-width: 20;
    
        /* Numeric Display */
        --dg-numeric-color: #000000;
        --dg-numeric-font-family: Verdana, Geneva, Tahoma, sans-serif;
        --dg-numeric-font-size: 2.5em;
    
        /* Subtitle */
        --dg-subtitle-color: #000000;
        --dg-subtitle-font-family: Verdana, Geneva, Tahoma, sans-serif;
        --dg-subtitle-font-size: 1em;
    }
    
    .dial-gauge {
        background-color: var(--dg-background-color, #ffffff);
        border: var(--dg-border, 1px solid #000000);
        border-radius: var(--dg-border-radius, 5px);
        box-shadow: var(--dg-box-shadow, 3px 3px 5px 1px rgba(0,0,0,0.35));
        display: var(--dg-display, inline-block);
        height: var(--dg-height, 200px);
        margin: var(--dg-margin, 0);
        text-align: center;
        width: var(--dg-width, 200px);
        z-index: var(--dg-z-index, 0);
    }

    #gauge-title {
        display: none;
        fill: var(--dg-title-color, #000000);
        font-family: var(--dg-title-font-family, Verdana, Geneva, Tahoma, sans-serif);
        font-size: var(--dg-title-font-size, 1.5em);
        text-anchor: middle;
    }

    svg > #gauge-background-arc {
        stroke: var(--dg-arc-background-color, #efefef);
        stroke-width: var(--dg-arc-width, 20);
    }

    svg > #gauge-arc {
        stroke: var(--dg-arc-color, #000000);
        stroke-width: var(--dg-arc-width, 20);
    }

    #gauge-numeric {
        display: none;
        dominant-baseline: middle;
        fill: var(--dg-numeric-color, #000000);
        font-family: var(--dg-numeric-font-family, Verdana, Geneva, Tahoma, sans-serif);
        font-size: var(--dg-numeric-font-size, 2.5em);
        text-anchor: middle;
    }

    #gauge-subtitle {
        display: none;
        fill: var(--dg-subtitle-color, #000000);
        font-family: var(--dg-subtitle-font-family, Verdana, Geneva, Tahoma, sans-serif);
        font-size: var(--dg-subtitle-font-size, 1.0em);
        text-anchor: middle;
    }
`);

/**
 * Creates a new DialGauge custom HTML element. Utilized in HTML as the <dial-gauge> tag.
 * @name DialGauge
 * @class
 * @extends HTMLElement
 */
export class DialGauge extends HTMLElement {
    /**
     * @description Specify observed attributes so that attributeChangedCallback will work.
     * @return {String[]} String array of attributes to be observed.
     */
    static get observedAttributes() {
        return [
            'value',
            'main-title',
            'sub-title',
            'scale-start',
            'scale-end',
            'scale-offset',
        ];
    }

    /**
     * @description Create a DialGauge element. Fires when an instance of
     * the element is created or updated.
     */
    constructor() {
        super();
        this.attachShadow({mode: 'open'});
        this.shadowRoot.adoptedStyleSheets = [styleSheet];
        this.shadowRoot.appendChild(template.content.cloneNode(true));
    }

    /**
     * @description Get the DialGauge element value.
     * @return {string} The value.
     */
    get value() {
        return this.getAttribute('value');
    }

    /**
     * @description Set the DialGauge element value.
     * @param {string} newValue The new value of the element.
     */
    set value(newValue) {
        this.setAttribute('value', newValue);
    }

    /**
     * @description Get the DialGauge element title.
     * @return {string} The title.
     */
    get mainTitle() {
        return this.getAttribute('main-title');
    }

    /**
     * @description Set the DialGauge element title.
     * @param {string} newValue The new title of the element.
     */
    set mainTitle(newValue) {
        this.setAttribute('main-title', newValue);
    }

    /**
     * @description Get the DialGauge element subtitle.
     * @return {string} The subtitle.
     */
    get subTitle() {
        return this.getAttribute('sub-title');
    }

    /**
     * @description Set the DialGauge element subtitle.
     * @param {string} newValue The new subtitle of the element.
     */
    set subTitle(newValue) {
        this.setAttribute('sub-title', newValue);
    }

    /**
     * @description Get the DialGauge element user scale starting value.
     * @return {float} The starting scale value.
     */
    get scaleStart() {
        return this.getAttribute('scale-start');
    }
    /**
     * @description Set the DialGauge element user scale starting value.
     * @param {float} newValue The new starting scale value.
     */
    set scaleStart(newValue) {
        this.setAttribute('scale-start', newValue);
    }

    /**
     * @description Get the DialGauge element user scale ending value.
     * @return {float} The ending scale value.
     */
    get scaleEnd() {
        return this.getAttribute('scale-end');
    }

    /**
     * @description Set the DialGauge element user scale ending value.
     * @param {float} newValue The new ending scale value.
     */
    set scaleEnd(newValue) {
        this.setAttribute('scale-end', newValue);
    }

    /**
     * @description Get the DialGauge user bottom offset angle.
     * @return {float} The bottom offset angle.
     */
    get scaleOffset() {
        return this.getAttribute('scale-offset');
    }

    /**
     * @description Set the DialGauge user bottom offset angle.
     * @param {float} newValue The new bottom offset angle.
     */
    set scaleOffset(newValue) {
        this.setAttribute('scale-offset', newValue);
    }

    /**
     * @name centerSVG
     * @description Calculate the center of the SVG container.
     * @return {Object}
     */
    centerSVG() {
        const container = this.shadowRoot.querySelector('.dial-gauge');
        return {
            x: container.clientWidth / 2,
            y: container.clientHeight / 2,
        };
    }

    /**
     * @name scaleDegrees
     * @description Scale user value to degrees.
     * @param {float} userValue The user numeric value.
     * @return {float}
     */
    scaleDegrees(userValue) {
        const start = this.scaleStart;
        const end = this.scaleEnd;
        return ((userValue - start) / (end - start)) * (360 - 2 * this.scaleOffset);
    }

    /**
     * @name polarToCartesian
     * @description Convert polar to cartesian coordinates.
     * @param {integer} centerX Arc center x coordinate.
     * @param {integer} centerY Arc center y coordinate.
     * @param {integer} radius Radius of the arc.
     * @param {integer} angleInDegrees Arc angle in degrees.
     * @return {Object}
     */
    polarToCartesian(centerX, centerY, radius, angleInDegrees) {
        const angleInRadians = (angleInDegrees - 90) * Math.PI / 180.0;
        return {
            x: centerX + (radius * Math.cos(angleInRadians)),
            y: centerY + (radius * Math.sin(angleInRadians)),
        };
    }

    /**
     * @name describeArc
     * @description Arc segment generator.
     * @param {integer} x Arc center x coordinate.
     * @param {integer} y Arc center y coordinate.
     * @param {integer} radius Radius of the arc.
     * @param {integer} startAngle Arc start angle in degrees.
     * @param {integer} endAngle Arc end angle in degrees.
     * @return {Object}
     */
    describeArc(x, y, radius, startAngle, endAngle) {
        endAngle = endAngle >= 180 ? 179.99 : endAngle;
        const start = this.polarToCartesian(x, y, radius, endAngle);
        const end = this.polarToCartesian(x, y, radius, startAngle);
        const largeArcFlag = endAngle - startAngle <= 180 ? '0' : '1';
        const d = [
            'M', start.x, start.y,
            'A', radius, radius, 0, largeArcFlag, 0, end.x, end.y,
        ].join(' ');
        return d;
    }

    /**
     * @name connectedCallback
     * @description Fires when an instance is inserted into the document.
     * @return {void}
     */
    connectedCallback() {
        // Display the main title if it is set.
        if (this.mainTitle !== '') {
            this.shadowRoot.querySelector('#gauge-title').style.display = 'initial';
        } else {
            this.shadowRoot.querySelector('#gauge-title').style.display = 'none';
        }
        // Display the sub title if it is set.
        if (this.mainTitle !== '') {
            this.shadowRoot.querySelector('#gauge-subtitle').style.display = 'initial';
        } else {
            this.shadowRoot.querySelector('#gauge-subtitle').style.display = 'none';
        }
        // Display the numeric and arc if value is set.
        if (this.value !== null) {
            this.shadowRoot.querySelector('#gauge-numeric').style.display = 'initial';
            this.shadowRoot.querySelector('#gauge-arc').style.display = 'initial';
        } else {
            this.shadowRoot.querySelector('#gauge-numeric').style.display = 'none';
            this.shadowRoot.querySelector('#gauge-arc').style.display = 'none';
        }
        // Render the background arc.
        const center = this.centerSVG();
        const arcRadius = this.shadowRoot.querySelector('.dial-gauge')
            .clientWidth * .30;
        this.shadowRoot.querySelector('#gauge-background-arc')
            .setAttribute(
                'd',
                this.describeArc(
                    center.x,
                    center.y,
                    arcRadius,
                    this.scaleOffset - 180,
                    180 - this.scaleOffset,
                ),
            );
    }

    /**
     * @name attributeChangedCallback
     * @description Fires when an attribute was added, removed, or updated.
     * @param {string} attrName The element attribute name.
     * @param {string} oldVal The old value of the attribute.
     * @param {string} newVal The new value of the attribute.
     * @return {void}
     */
    attributeChangedCallback(attrName, oldVal, newVal) {
        switch (attrName) {
        case 'value':
            if (newVal !== null) {
                // Check if value is within scale limits
                if (newVal >= parseFloat(this.scaleStart) &&
                    newVal <= parseFloat(this.scaleEnd)) { // Within scale limits.
                    // Value is set, display the arc and numeric.
                    this.shadowRoot.querySelector('#gauge-arc').style.display = 'initial';
                    this.shadowRoot.querySelector('#gauge-numeric').style.display = 'initial';
                    // Get the center of component.
                    const center = this.centerSVG();
                    // Calculate the arc radius based on component width.
                    const arcRadius = this.shadowRoot.querySelector('.dial-gauge')
                        .clientWidth * .30;
                    // Scale new user value to degree.
                    const scaledValue = this.scaleDegrees(newVal);
                    // Update the arc display.
                    this.shadowRoot.querySelector('#gauge-arc')
                        .setAttribute(
                            'd',
                            this.describeArc(
                                center.x,
                                center.y,
                                arcRadius,
                                this.scaleOffset - 180,
                                scaledValue - (180 - this.scaleOffset),
                            ),
                        );
                    // Update the numeric display.
                    this.shadowRoot.querySelector('#gauge-numeric').textContent = newVal;
                } else { // Outside scale limits.
                    this.shadowRoot.querySelector('#gauge-arc').style.display = 'none';
                    this.shadowRoot.querySelector('#gauge-numeric').textContent = 'OL';
                }
            } else {
                // Value is not set, hide the arc and numeric.
                this.shadowRoot.querySelector('#gauge-arc').style.display = 'none';
                this.shadowRoot.querySelector('#gauge-numeric').style.display = 'none';
            }
            break;
        case 'main-title':
            if (newVal !== oldVal) {
                // Update element title attribute.
                this.shadowRoot.querySelector('#gauge-title').textContent = newVal;
            }
            break;
        case 'sub-title':
            if (newVal !== oldVal) {
                // Update element subtitle attribute.
                this.shadowRoot.querySelector('#gauge-subtitle').textContent = newVal;
            }
            break;
        case 'scale-start':
            if (newVal !== null && newVal !== oldVal) {
                // Check for scale start out of range.
                if (this.value !== null &&
                    newVal <= parseFloat(this.value)) { // Scale start is in range.
                    // Get the center of component.
                    const center = this.centerSVG();
                    // Calculate the arc radius based on component width.
                    const arcRadius = this.shadowRoot.querySelector('.dial-gauge')
                        .clientWidth * .30;
                    // Scale the user value to degree.
                    const scaledValue = this.scaleDegrees(this.value);
                    // Re-initialize the arc and numeric displays.
                    this.shadowRoot.querySelector('#gauge-arc').style.display = 'initial';
                    this.shadowRoot.querySelector('#gauge-numeric').textContent = this.value;
                    // Update the arc display.
                    this.shadowRoot.querySelector('#gauge-arc')
                        .setAttribute(
                            'd',
                            this.describeArc(
                                center.x,
                                center.y,
                                arcRadius,
                                this.scaleOffset - 180,
                                scaledValue - (180 - this.scaleOffset),
                            ),
                        );
                } else { // Scale start is out of range.
                    this.shadowRoot.querySelector('#gauge-arc').style.display = 'none';
                    this.shadowRoot.querySelector('#gauge-numeric').textContent = 'OL';
                }
            }
            break;
        case 'scale-end':
            if (newVal !== null && newVal !== oldVal) {
                // Check for scale end out of range.
                if (this.value !== null &&
                    newVal >= parseFloat(this.value)) { // Scale end is in range.
                    // Get the center of component.
                    const center = this.centerSVG();
                    // Calculate the arc radius based on component width.
                    const arcRadius = this.shadowRoot.querySelector('.dial-gauge')
                        .clientWidth * .30;
                    // Scale the user value to degree.
                    const scaledValue = this.scaleDegrees(this.value);
                    // Re-initialize the arc and numeric displays.
                    this.shadowRoot.querySelector('#gauge-arc').style.display = 'initial';
                    this.shadowRoot.querySelector('#gauge-numeric').textContent = this.value;
                    // Update the arc display.
                    this.shadowRoot.querySelector('#gauge-arc')
                        .setAttribute(
                            'd',
                            this.describeArc(
                                center.x,
                                center.y,
                                arcRadius,
                                this.scaleOffset - 180,
                                scaledValue - (180 - this.scaleOffset),
                            ),
                        );
                } else { // Scale end is out of range.
                    this.shadowRoot.querySelector('#gauge-arc').style.display = 'none';
                    this.shadowRoot.querySelector('#gauge-numeric').textContent = 'OL';
                }
            }
            break;
        case 'scale-offset':
            if (newVal !== null && newVal !== oldVal) {
                // Check for offset greater than 180 degrees.
                this.scaleOffset = this.scaleOffset > 180 ? 180 : this.scaleOffset;
                // Get the center of component.
                const center = this.centerSVG();
                // Calculate the arc radius based on component width.
                const arcRadius = this.shadowRoot.querySelector('.dial-gauge')
                    .clientWidth * .30;
                // Scale the user value to degree.
                const scaledValue = this.scaleDegrees(this.value);
                // Update the background arc.
                this.shadowRoot.querySelector('#gauge-background-arc')
                    .setAttribute(
                        'd',
                        this.describeArc(
                            center.x,
                            center.y,
                            arcRadius,
                            this.scaleOffset - 180,
                            180 - this.scaleOffset,
                        ),
                    );
                // Update the arc display.
                this.shadowRoot.querySelector('#gauge-arc')
                    .setAttribute(
                        'd',
                        this.describeArc(
                            center.x,
                            center.y,
                            arcRadius,
                            this.scaleOffset - 180,
                            scaledValue - (180 - this.scaleOffset),
                        ),
                    );
            }
            break;
        }
    }
}

customElements.define('dial-gauge', DialGauge);
